### ISSUE TITLE
b-carousel-slide :img-src lazy-loading

### ISSUE THREAD

**Issue by drzraf** [bootstrap-vue/bootstrap-vue#1210]
I would like to have my (dozens+) carousels images being lazy loaded rather than blocking rendering.
I don't know if it's impossible or just undocumented (there is no `*lazy*` on [carousel documentation](https://bootstrap-vue.js.org/docs/components/carousel/))
Any hint welcome.

**Comment by tmorehouse** [comment]
You can specify your own image component inside the slides.  Just use b-img-lazy instead of img or b-img

**Comment by drzraf** [comment]
I can't find a trace of `<b-carousel-slide>` in the sourcecode.
Documentation says that:

> Internally, `<b-carousel-slide>` uses the `<b-img>` component to render the images. The img-* props map to the corresponsing props available to `<b-img>`.

But this does not precisely describe what is it useful for and/or what advantage does it provide over `<b-img>`.
How should I wrap a `<b-img-lazy>` in order to stick with default behavior of built-in `<b-carousel-slide>`?

**Comment by tmorehouse** [comment]
B-Carousel-slide code is at https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/lib/components/carousel-slide.vue

```html
      <b-carousel-slide>
        <img slot="img" class="d-block img-fluid w-100" width="1024" height="480"
             src="https://lorempixel.com/1024/480/technics/5/" alt="image slot">
      </b-carousel-slide>
```

You can just replace `<img ...>` above with `<b-img-lazy ...>`  but you should set the width and height of the original image on b-img-lazy so that the carousel will be sized correctly (as carousel itself doesn't have a defined siz, and if a size for image is set, it only applies to the image specified via the img-src props.  All of your slide images should be the same dimensions, and should have the fluid or fluid-grop prop set.

**Comment by drzraf** [comment]
Good tip! I had overlooked the sample about `technics/5/` where `<slot>` is used.
**But** when using this `<b-carousel-slide> <img /> </b-carousel-slide>` nested-syntax, the <img/> is nested *inside* the `div.carousel-caption` the consequence is that the image is not visible.

I finally worked around this by hardcoding the template:
```html
<!-- first image is not lazy-loaded: use built-in b-carousel-slide -->
<b-carousel-slide :key="foo.id + 'thumb0'" width="250" height="250" :img-src="foo.thumbnails[0]" />

<!-- because wrapping inside a <b-carousel-slide> does not work: hardcode the expanded template -->
<div class="carousel-item" role="listitem" v-for="(t, id) in foo.thumbnails.slice(1)" :key="foo.id + 'thumb' + id + 1">
  <b-img-lazy fluid-grow block width="250" height="250" :src="t" />
</div>
```

... **but** encountered another issue related to `<b-img-lazy/>`: if the image can't be loaded in the `interval` seconds, it'll never show-up (and will never load afterward)

**Comment by tmorehouse** [comment]
b-img-lazy checks to see if the image (the placeholder img) is _visible_ in the viewport, and probably in this case, the image wont be triggered until it starts to slide into view (there are never more than two slides with display:block at any one time (the other sides would have visibility none set based on BS V4 CSS).

As well, b-img-lazy is trigger on scrolling (i.e. window.scroll event).  A sliding carousel normally wouldn't trigger a window scroll.

**Comment by tvld** [comment]
Seems that this works fine, no need for ``setInterval()``:
```js
    methods: {
      onSlideStart (slide) {
        this.sliding = true
        // for laxy loading, we invoke scroll event (ignore Lint Error... it seem to work ok)
        /* global Event */
        window.dispatchEvent(new Event("scroll"))
      },
```

**Comment by tmorehouse** [comment]
Note that this also assumes that the time to load the image over the network is faster than the slide interval.

**Comment by tvld** [comment]
That is right indeed. Still, with a slow network, rsts what can happen, is that user needs to wait one slide interval... so if that's not to big, it will be ok. To load first slide, so that separate showing of ``b-carousel-slide`` is no longer needed and we can load all slides lazy, in main js I have:
```js
  // start scroll on load the pager
  setTimeout(function () {
    /* global Event */
    window.dispatchEvent(new Event("scroll"))
  }, 
```

**Comment by tmorehouse** [comment]
One option, might be for us to have `b-img-lazy` listen for global ($root) carousel slide events, and use that to also trigger the visibility check (in a $nextTick), but that would definitely cause a lot of overhead calculations for this edge case.

**Comment by tvld** [comment]
Yeah ... that could be. But indeed it is a bit out of logic to let child  ``b-im-lazy``, listen to one potential parent ``carousel``...

What about letting carousel always invoke a scroll event on each slide action? What is the penalty for that? After all ... sliding is also scrolling, so that does not seem to be too unconventional...

**Comment by tmorehouse** [comment]
Technically it is "transitioning" (CSS 3d transform) 😉 

There are a few other components that listen to scroll events, so I guess it would be determined by the number of those components on the page.

**Comment by tmorehouse** [comment]
And `new Event()` isn't supported in IE 10/11 https://developer.mozilla.org/en-US/docs/Web/API/Event/Event#Browser_compatibility

**Comment by tvld** [comment]
Hmm, IE10/11 is still being used...So we still have no elegant solution?
I have this now and was so happy with it...  ;-P

```js
  export default {
    name: 'app',
    mounted: function () {
      this.$nextTick(function () {
        // Code that will run only after the entire view has been re-rendered
        console.log("Force lazy loading event.")
        /* global Event */
        window.dispatchEvent(new Event("scroll"))
      })
    }
```

**Comment by tmorehouse** [comment]
Lots of large corporations are still using Windows 7, which ships with IE 11.

**Comment by tmorehouse** [comment]
You would need to polyfill for that: https://stackoverflow.com/a/26596324

or use the deprecated `document.createEvent(..)` https://stackoverflow.com/questions/27176983/dispatchevent-not-working-in-ie11 (and use browser feature sniffing)

**Comment by tvld** [comment]
``npm install custom-event-polyfill``, seems to work so far )

**Comment by blaskovicz** [comment]
Why was this closed? Shouldn't the `b-lazy-img` component be updated to have whatever logic it needs to render correctly under most circumstances?

**Comment by blaskovicz** [comment]
https://github.com/blaskovicz/house-rank/commit/0118fc398719f3db65d3d91be085536ff122dbfe
